src/tree.rs

```diff
@@ -102,42 +135,116 @@ fn cond_swap(swap: bool, node: pallas::Base, sibling: pallas::Base) -> Pair {
     // MerkleCRH Sinsemilla hash domain.
     let domain = HashDomain::new(MERKLE_CRH_PERSONALIZATION);
 
-    domain
-        .hash(
-            iter::empty()
-                .chain(i2lebsp_k(l_star).iter().copied())
-                .chain(
-                    pair.left
-                        .to_le_bits()
-                        .iter()
-                        .by_val()
-                        .take(L_ORCHARD_MERKLE),
-                )
-                .chain(
-                    pair.right
-                        .to_le_bits()
-                        .iter()
-                        .by_val()
-                        .take(L_ORCHARD_MERKLE),
-                ),
-        )
-        .unwrap()
+    domain.hash(
+        iter::empty()
+            .chain(i2lebsp_k(l).iter().copied())
+            .chain(
+                pair.left
+                    .to_le_bits()
+                    .iter()
+                    .by_val()
+                    .take(L_ORCHARD_MERKLE),
+            )
+            .chain(
+                pair.right
+                    .to_le_bits()
+                    .iter()
+                    .by_val()
+                    .take(L_ORCHARD_MERKLE),
+            ),
+    )
+}
+
+/// A newtype wrapper for leaves and internal nodes in the Orchard
+/// incremental note commitment tree.
+#[derive(Clone, Debug)]
+pub struct OrchardIncrementalTreeDigest(CtOption<pallas::Base>);
+
+impl OrchardIncrementalTreeDigest {
+    /// Creates an incremental tree life digest for the specified
+    /// Pallas base field element.
+    pub fn from_cmx(value: &ExtractedNoteCommitment) -> Self {
+        OrchardIncrementalTreeDigest(CtOption::new(**value, 1.into()))
+    }
+
+    /// Convert this digest to its canonical byte representation.
+    pub fn to_bytes(&self) -> Option<[u8; 32]> {
+        <Option<pallas::Base>>::from(self.0).map(|b| b.to_bytes())
+    }
+
+    /// Parses a incremental tree leaf digest from the bytes of
+    /// a note commitment.
+    pub fn from_bytes(bytes: &[u8; 32]) -> Self {
+        OrchardIncrementalTreeDigest(pallas::Base::from_bytes(bytes))
+    }
+}
+
+/// This instance is should only be used for hashtable key comparisons.
+impl std::cmp::PartialEq for OrchardIncrementalTreeDigest {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.ct_eq(&other.0).into()
+    }
+}
+
+/// This instance is should only be used for hashtable key comparisons.
+impl std::cmp::Eq for OrchardIncrementalTreeDigest {}
+
+/// This instance is should only be used for hashtable key comparisons.
+impl std::hash::Hash for OrchardIncrementalTreeDigest {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        <Option<pallas::Base>>::from(self.0)
+            .map(|b| b.to_bytes())
+            .hash(state)
+    }
+}
+
+impl Hashable for OrchardIncrementalTreeDigest {
+    fn empty_leaf() -> Self {
+        OrchardIncrementalTreeDigest(CtOption::new(pallas::Base::from_u64(2), 1.into()))
```

Comment:
`pallas::Base::from_u64(2)` is repeated in several places; please create a named constant.